### PR TITLE
STL-1788 | Added session logic with redis client | Update some …

### PR DESCRIPTION
**Pull request by VictorDelCampo**
# Short Description
- Updated session logic to use the new redis client implementation

# Changes
- get/save(override) methods using the current_user as a part of the key identifier
- deserialize_session_data because the redis client implementation is returning a string
- already updated two classes in test_session_data so they are pure pytest

# Feedback
- I have a big open ear :)

# How to review this Pull Request
[Link to Confluence](https://digitalservice4germany.atlassian.net/wiki/spaces/STL/pages/117637157/Development+Process#Approve-or-comment) (internal)

### CODE REVIEW

File: webapp/app/forms/session_data.py [hunk 1]
```diff
@@ -45,4 +52,9 @@ def deserialize_session_data(serialized_session, ttl: Optional[int] = None):
 
 
 def override_session_data(stored_data, session_data_identifier='form_data'):
-    session[session_data_identifier] = serialize_session_data(stored_data)
+    key = create_key_identifier_with_user_id(session_data_identifier)
+    FormDataController().save_to_redis(key, serialize_session_data(stored_data))
```
Comment (JulianRoesner) [inline webapp/app/forms/session_data.py:56]: In my head, I was thinking of sth like that to keep the functionality of having different types of session_data in one session:
```suggestion
    key = create_key_identifier_with_user_id(session_data_identifier)
    old_session_data = deserialize(FormDataController().get_from_redis(key))
    old_session_data[session_data_identifier] = stored_data
    FormDataController().save_to_redis(key, serialize_session_data(old_session_data))
```

However, I think I like the approach of creating the identifier in the way you do it even more as we use the redis directly 👍

File: webapp/app/routes.py [hunk 2]
```diff
@@ -377,7 +377,7 @@ def register_testing_request_handlers(app):
         return
 
     @csrf.exempt
-    @app.route('/testing/set_data/<session_identifier>', methods=['POST'])
+    @app.route('/testing/set_session_data/<session_identifier>', methods=['POST'])
```
Comment (JulianRoesner) [inline webapp/app/routes.py:380]: Renamed that route to reflect that we only set data in the session

File: webapp/tests/conftest.py [hunk 3]
```diff
@@ -36,21 +33,52 @@ def test_request_context(app):
...
-            if form_data:
-                req.request.data = ImmutableMultiDict(form_data)
-                req.request.form = ImmutableMultiDict(form_data)
+                FormDataController().save_to_redis(CURRENT_USER_IDNR + '_' + session_identifier,
+                                                   serialize_session_data(stored_data))
+            else:
+                FormDataController().save_to_redis(CURRENT_USER_IDNR + '_' + session_identifier,
+                                                   serialize_session_data(form_data))
             yield req
+
```
Comment (JulianRoesner) [inline webapp/tests/conftest.py:68]: As we do not set the data on the request context, I am wondering if we could use a second fixture (s. below). This would allow us to keep the functionality of the test_request_context the same which we need for steps where we do not use the redis session:
```
@pytest.fixture
 def new_test_request_context_with_data_in_session(app, new_test_request_context):
     @contextmanager
     def _new_test_request_context(method='GET', form_data=None, session_data=None, session_identifier='form_data'):
          req =  new_test_request_context(method=method, form_data)
          if session_data:
                  FormDataController().save_to_redis(CURRENT_USER_IDNR + '_' + session_identifier,
                                                     serialize_session_data(stored_data))
          yield req
```
As this should still create a correct post request, we need to set the form data not in this method but set them as part of the request_context. form_data is the data that the post request sends while stored_data is the data that we already have in the session (in the cookie context it is part of the cookie, in the session context it is part of the session)

File: webapp/tests/conftest.py [hunk 4]
```diff
@@ -36,21 +33,52 @@ def test_request_context(app):
...
     return _new_test_request_context
 
 
+@pytest.fixture
+def test_request_context_with_person_b_disability(app):
```
Comment (JulianRoesner) [inline webapp/tests/conftest.py:73]: I would name it `test_post_request_context...` if it is always a post request context. Wdyt?

File: webapp/tests/conftest.py [hunk 5]
```diff
@@ -36,21 +33,52 @@ def test_request_context(app):
...
+    from app.forms.session_data import serialize_session_data
+    from app.data_access.form_data_controller import FormDataController
+    with app.test_request_context(method="POST") as req:
+        FormDataController().save_to_redis(CURRENT_USER_IDNR + '_form_data',
+                                            serialize_session_data({'person_b_has_disability': 'yes'}))
+        yield req
```
Comment (JulianRoesner) [inline webapp/tests/conftest.py:79]: We could use the fixture from above. wdyt?

File: webapp/tests/forms/steps/lotse/test_pauschbetrag.py [hunk 6]
```diff
@@ -229,10 +229,10 @@ def test_if_merkzeichen_given_then_get_pauschbetrag_returns_result_of_calculate_
             'person_a_disability_degree': 25,
             'person_a_has_merkzeichen_bl': True,
             'person_a_has_merkzeichen_tbl': True,
-            'person_a_has_merkzeichen_h': True
+            'person_a_has_merkzeichen_h': True,
+            'person_a_requests_pauschbetrag': 'yes',
         })
         form_data = MultiDict({
-            'person_a_requests_pauschbetrag': 'yes',
```
Comment (JulianRoesner) [inline webapp/tests/forms/steps/lotse/test_pauschbetrag.py:235]: Why did you change that? That is for all these tests? I think this might have to do with the issue that we do not set the post data in the test_request_context. Maybe let's revisit these tests

Comment (JulianRoesner) [reply webapp/tests/forms/steps/lotse/test_pauschbetrag.py:235]: that is for all the tests in this file

File: webapp/tests/forms/steps/lotse/test_personal_data_steps.py [hunk 7]
```diff
@@ -101,9 +101,7 @@ def test_if_multiple_users_then_show_multiple_text(self, app):
         expected_request_new_tax_number_label = ngettext('form.lotse.steuernummer.request_new_tax_number',
                                                          'form.lotse.steuernummer.request_new_tax_number',
                                                          num=expected_number_of_users)
-        with app.test_request_context(method='GET') as req:
-            req.session = SecureCookieSession(
-                {_LOTSE_DATA_KEY: create_session_form_data(session_data)})
+        with new_test_request_context(method='GET', form_data=session_data):
```
Comment (JulianRoesner) [inline webapp/tests/forms/steps/lotse/test_personal_data_steps.py:104]: We probably want to set it to stored data not to form_data here, too?

File: webapp/tests/forms/steps/lotse/test_personal_data_steps.py [hunk 8]
```diff
@@ -264,9 +262,7 @@ def test_if_multiple_users_then_show_multiple_text(self, app):
         expected_step_intro = _(
             'form.lotse.person-a-intro') if expected_number_of_users > 1 else None
 
-        with app.test_request_context(method='GET') as req:
-            req.session = SecureCookieSession(
-                {_LOTSE_DATA_KEY: create_session_form_data(session_data)})
+        with new_test_request_context(method='GET', form_data=session_data) as req:
```
Comment (JulianRoesner) [inline webapp/tests/forms/steps/lotse/test_personal_data_steps.py:265]: We probably want to set it to stored data not to form_data here, too?

File: webapp/tests/forms/test_session_data.py [hunk 9]
```diff
@@ -40,15 +25,15 @@ def test_if_session_data_and_default_data_different_then_update_session_data(sel
...
+    def test_if_session_data_for_the_current_user_is_returned(self, testing_current_user):
         form_data = {"brother": "Luigi"}
         other_form_data = {"enemy": "Bowser"}
         expected_data = {**other_form_data}
         override_session_data(form_data)
-        testing_database.setattr(CURRENT_USER, MagicMock(idnr_hashed="9876543210"))
-        override_session_data(other_form_data)
-        session_data = get_session_data('form_data')
-        assert expected_data == session_data
+        with patch(CURRENT_USER, MagicMock(idnr_hashed="9876543210")):
```
Comment (JulianRoesner) [inline webapp/tests/forms/test_session_data.py:33]: Why do we need that if we already use the fixture?

File: webapp/tests/forms/test_session_data.py [hunk 10]
```diff
@@ -40,15 +25,15 @@ def test_if_session_data_and_default_data_different_then_update_session_data(sel
         session_data = get_session_data('form_data', default_data=default_data)
         assert expected_data == session_data
 
-    def test_if_session_data_for_the_current_user_is_returned(self, testing_database):
+    def test_if_session_data_for_the_current_user_is_returned(self, testing_current_user):
         form_data = {"brother": "Luigi"}
         other_form_data = {"enemy": "Bowser"}
         expected_data = {**other_form_data}
         override_session_data(form_data)
```
Comment (JulianRoesner) [inline webapp/tests/forms/test_session_data.py:32]: couldn't we set the data directly to the database?

File: webapp/tests/forms/test_session_data.py [hunk 11]
```diff
@@ -40,15 +25,15 @@ def test_if_session_data_and_default_data_different_then_update_session_data(sel
         session_data = get_session_data('form_data', default_data=default_data)
         assert expected_data == session_data
 
-    def test_if_session_data_for_the_current_user_is_returned(self, testing_database):
+    def test_if_session_data_for_the_current_user_is_returned(self, testing_current_user):
```
Comment (JulianRoesner) [inline webapp/tests/forms/test_session_data.py:28]: I think we should keep the test for an incorrect identifier, because it is different from the current_user test. wdyt?

File: webapp/tests/conftest.py (hunk 4 above)
Comment (VictorDelCampo) [reply webapp/tests/conftest.py:73]: I just moved this fixture from an old file to the conftest file, but yes I can rename it

File: webapp/tests/forms/test_session_data.py (hunk 9 above)
Comment (VictorDelCampo) [reply webapp/tests/forms/test_session_data.py:36]: Because we are changing the current user, so we create another entry in redis different than the previous user

File: webapp/tests/conftest.py (hunk 5 above)
Comment (VictorDelCampo) [reply webapp/tests/conftest.py]: Done

File: webapp/tests/forms/steps/lotse/test_pauschbetrag.py (hunk 6 above)
Comment (VictorDelCampo) [reply webapp/tests/forms/steps/lotse/test_pauschbetrag.py]: Reverted and added new fixture

File: webapp/tests/forms/test_session_data.py (hunk 11 above)
Comment (VictorDelCampo) [reply webapp/tests/forms/test_session_data.py:31]: I will readd it!

File: webapp/tests/forms/test_session_data.py (hunk 10 above)
Comment (JulianRoesner) [reply webapp/tests/forms/test_session_data.py:34]: Let's keep it that way

File: webapp/tests/forms/test_session_data.py (hunk 11 above)
Comment (JulianRoesner) [reply webapp/tests/forms/test_session_data.py:30]: Thanks

**Review by punknoir101 — COMMENTED** [review]
Just some proposals regarding the naming for the new session/redis implementations:
form_data_controller --> redis_connector_service 
session_data --> user_scoped_redis_store 
cookie_data --> session_data (We are not using "real" cookie functions, but session data with storage cookie.)

**Comment by SannyNguyenHung** [comment]
Implemented:
* SessionStorage <- All data stored in redis
* CookieStorage <- All data stored within the cookie
* FormStorage <- Base class for both

The StepChooser has a new parameter: form_storage. The StepChooser does not decide which implementation is taken but the implementer. The implementer specifies the corresponding storage (EligibilityStepChooser -> CookieStorage, Lotse -> SessionStorage, ...).

File: webapp/app/data_access/storage/cookie_storage.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
+from typing import Optional
+from flask import session
+
+from app.data_access.storage.form_storage import FormStorage
+
+
+class CookieStorage(FormStorage):
```
Comment (VictorDelCampo) [inline webapp/app/data_access/storage/cookie_storage.py:7]: Shouldn't we add the method docs in the super-class FormStorage? And remove them from the sub-class CookieStorage? The other sub-class SessionStorage doesn't have any docs

File: webapp/app/data_access/storage/session_storage.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
+from typing import Optional
+
+from pydantic import MissingError
+from flask_login import current_user
+
+from app.data_access.form_data_controller import FormDataController
+from app.data_access.storage.form_storage import FormStorage
+
+
+class SessionStorage(FormStorage):
```
Comment (VictorDelCampo) [inline webapp/app/data_access/storage/session_storage.py:10]: We could maybe make both SessionStorage and CookieStorage be of singleton use so that everytime we need them they are not newly instantiated? wdyt?

File: webapp/app/data_access/storage/cookie_storage.py (hunk 12 above)
Comment (SannyNguyenHung) [reply webapp/app/data_access/storage/cookie_storage.py:7]: Done

File: webapp/app/data_access/storage/session_storage.py (hunk 13 above)
Comment (SannyNguyenHung) [reply webapp/app/data_access/storage/session_storage.py:10]: Not necessary :) why should we need a singleton?

Comment (VictorDelCampo) [reply webapp/app/data_access/storage/session_storage.py:10]: I like singleton :) but if you think is not needed, then it is ok how it is implemented

**Comment by punknoir101** [comment]
I have seen some formatting issues only (e.g. lines between methods)

File: webapp/app/data_access/storage/form_storage.py [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        Gets data from a storage that is associated with the currect session. The data is return deserialized.
+
+        :param data_identifier: A string used to identify what data should be taken from the cookie
+        :param ttl: The time to live for the cookie
+        :param default_data: Default data that will be used to replace missing data points
+        """
+        pass
+
+    @abstractmethod
+    def override_data(self, data_to_store, data_identifier):
```
Comment (JulianRoesner) [inline webapp/app/data_access/storage/form_storage.py:28]: Why don't we use static methods or class methods for these, too? We would not need to initialize the class in the steps

File: webapp/app/forms/steps/eligibility_steps.py [hunk 15]
```diff
@@ -161,18 +160,19 @@ def __init__(self, stored_data=None, render_info=None, *args, **kwargs):
...
         # Remove all eligibility data as the flow is restarting
         stored_data = {}
-        self._override_session_data(stored_data, session_data_identifier=self.session_data_identifier)
+        self._override_storage_data(stored_data, data_identifier=self.session_data_identifier)
         self.render_info.additional_info['next_button_label'] = _('form.eligibility.check-now-button')
 
-    def _override_session_data(self, stored_data, session_data_identifier=None):
-        override_data_in_cookie(stored_data, session_data_identifier)
+    def _override_storage_data(self, stored_data, data_identifier=None):
+        CookieStorage().override_data(stored_data, data_identifier)
```
Comment (JulianRoesner) [inline webapp/app/forms/steps/eligibility_steps.py:175]: Why don't we use `self.form_storage.override_data` here instead of the_override_storage_data method?

Comment (JulianRoesner) [inline webapp/app/forms/steps/eligibility_steps.py:175]:
```suggestion
        self.form_storage.override_data(stored_data, data_identifier)
```
WDYT?

Comment (JulianRoesner) [reply webapp/app/forms/steps/eligibility_steps.py:175]: Or is this because of the implementation in the steuerlotse_step.py.

File: webapp/app/forms/steps/steuerlotse_step.py [hunk 16]
```diff
@@ -77,10 +78,10 @@ def _post_handle(self):
...
         return self.render()
 
-    def _override_session_data(self, stored_data, session_data_identifier=None):
-        if session_data_identifier is None:
-            session_data_identifier = self.session_data_identifier
-        override_session_data(stored_data, session_data_identifier)
+    def _override_storage_data(self, stored_data, data_identifier=None):
+        if data_identifier is None:
+            data_identifier = self.data_identifier
+        self.form_storage.override_data(stored_data, data_identifier)
```
Comment (JulianRoesner) [inline webapp/app/forms/steps/steuerlotse_step.py:84]: S. question for eligibility_step

File: webapp/tests/forms/test_cookie_data.py [hunk 17]
```diff
@@ -4,8 +4,8 @@
 import pytest
```
Comment (JulianRoesner) [inline webapp/tests/forms/test_cookie_data.py:4]: Probably we should rename the file to test_cookie_storage.py, wdyt?

File: webapp/tests/forms/test_session_data.py [hunk 18]
```diff
@@ -1,122 +1,159 @@
 import datetime
```
Comment (JulianRoesner) [inline webapp/tests/forms/test_session_data.py:1]: Probaly we should also rename that file to test_session_storage.py, wdyt?

**Review by JulianRoesner — APPROVED** [review]
Just some comments about the steps. Other than that it lgtm

File: webapp/app/data_access/storage/form_storage.py (hunk 14 above)
Comment (SannyNguyenHung) [reply webapp/app/data_access/storage/form_storage.py]: Done

File: webapp/app/forms/steps/eligibility_steps.py (hunk 15 above)
Comment (SannyNguyenHung) [reply webapp/app/forms/steps/eligibility_steps.py]: done -> we need it in the steuerlotse_step because EligibilityStartDisplaySteuerlotseStep needs that functionality.

File: webapp/app/forms/steps/steuerlotse_step.py (hunk 16 above)
Comment (SannyNguyenHung) [reply webapp/app/forms/steps/steuerlotse_step.py]: done  -> we need it in the steuerlotse_step because EligibilityStartDisplaySteuerlotseStep needs that functionality.

File: webapp/tests/forms/test_cookie_data.py (hunk 17 above)
Comment (SannyNguyenHung) [reply webapp/tests/forms/test_cookie_data.py:4]: done!

**Review by SannyNguyenHung — APPROVED** [review]
LGTM!
